syntax = "proto3";

package layout;

// Сервис для расчета укладки графа
service LayoutService {
  // Основной метод для расчета укладки графа
  rpc CalculateLayout(LayoutRequest) returns (LayoutResponse);
  
  // Проверка здоровья сервиса
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// === ВХОДНЫЕ ДАННЫЕ ===

// Запрос на расчет укладки
message LayoutRequest {
  repeated BlockInput blocks = 1;
  repeated LinkInput links = 2;
  LayoutOptions options = 3;
}

// Блок для входных данных (без координат)
message BlockInput {
  string id = 1;
  string content = 2;
  bool is_pinned = 3;              // Закреплен ли блок за уровнем
  map<string, string> metadata = 4; // Дополнительные данные
}

// Связь между блоками
message LinkInput {
  string id = 1;
  string source_id = 2;
  string target_id = 3;
  map<string, string> metadata = 4; // Дополнительные данные
}

// Опции для алгоритма укладки
message LayoutOptions {
  int32 max_layers = 1;           // Максимальное количество слоев (по умолчанию автоматически)
  int32 max_levels = 2;           // Максимальное количество уровней (по умолчанию автоматически)
  int32 blocks_per_sublevel = 3;  // Блоков на подуровень (по умолчанию автоматически)
  bool optimize_layout = 4;       // Применять двухпроходную оптимизацию (по умолчанию true)
  int32 sublevel_spacing = 5;     // Расстояние между подуровнями в пикселях (по умолчанию 200)
  int32 layer_spacing = 6;        // Расстояние между слоями в пикселях (по умолчанию 250)
}

// === ВЫХОДНЫЕ ДАННЫЕ ===

// Ответ с результатом укладки
message LayoutResponse {
  repeated BlockWithPosition blocks = 1;
  repeated Level levels = 2;
  repeated Sublevel sublevels = 3;
  LayoutStatistics statistics = 4;
  bool success = 5;
  string error_message = 6;
}

// Блок с позицией после укладки
message BlockWithPosition {
  string id = 1;
  string content = 2;
  float x = 3;                    // X координата в пикселях
  float y = 4;                    // Y координата в пикселях
  int32 layer = 5;                // Слой (временная последовательность)
  int32 level = 6;                // Уровень (группировка)
  int32 sublevel_id = 7;          // ID подуровня к которому принадлежит блок
  bool is_pinned = 8;             // Закреплен ли блок за уровнем
  map<string, string> metadata = 9; // Дополнительные данные
}

// Уровень (группирует несколько подуровней)
message Level {
  int32 id = 1;
  repeated int32 sublevel_ids = 2; // ID подуровней в этом уровне
  float min_x = 3;                 // Левая граница уровня
  float max_x = 4;                 // Правая граница уровня
  float min_y = 5;                 // Верхняя граница уровня
  float max_y = 6;                 // Нижняя граница уровня
  int32 color = 7;                 // Цвет уровня (RGB в виде int)
  string name = 8;                 // Название уровня
}

// Подуровень (горизонтальная дорожка для блоков)
message Sublevel {
  int32 id = 1;
  float y = 2;                     // Y координата центра подуровня
  repeated string block_ids = 3;   // ID блоков в этом подуровне
  float min_x = 4;                 // Левая граница подуровня
  float max_x = 5;                 // Правая граница подуровня
  int32 color = 6;                 // Цвет подуровня (RGB в виде int)
  int32 level_id = 7;              // ID уровня к которому принадлежит подуровень
  float height = 8;                // Высота подуровня в пикселях
}

// Статистика результата укладки
message LayoutStatistics {
  int32 total_blocks = 1;          // Общее количество блоков
  int32 total_links = 2;           // Общее количество связей
  int32 total_levels = 3;          // Количество уровней
  int32 total_sublevels = 4;       // Количество подуровней
  int32 max_layer = 5;             // Максимальный слой
  float total_width = 6;           // Общая ширина графа
  float total_height = 7;          // Общая высота графа
  int32 processing_time_ms = 8;    // Время обработки в миллисекундах
  bool is_acyclic = 9;             // Является ли граф ациклическим
  int32 isolated_blocks = 10;      // Количество изолированных блоков
}

// === ВСПОМОГАТЕЛЬНЫЕ СООБЩЕНИЯ ===

// Запрос проверки здоровья
message HealthCheckRequest {
  string service = 1;
}

// Ответ проверки здоровья
message HealthCheckResponse {
  enum ServingStatus {
    UNKNOWN = 0;
    SERVING = 1;
    NOT_SERVING = 2;
    SERVICE_UNKNOWN = 3;
  }
  ServingStatus status = 1;
  string message = 2;
} 